src/pytorch_lightning/strategies/strategy.py

```diff
@@ -327,49 +335,61 @@ def load_optimizer_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
             optimizer.load_state_dict(opt_state)
             optimizer_to_device(optimizer, self.root_device)
 
-    def training_step(self, *args, **kwargs) -> STEP_OUTPUT:
+    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
         """The actual training step.
 
         See :meth:`~pytorch_lightning.core.module.LightningModule.training_step` for more details
         """
         with self.precision_plugin.train_step_context():
+            assert isinstance(self.model, pl.LightningModule) or isinstance(
+                self.model, _LightningPrecisionModuleWrapperBase
+            )
             return self.model.training_step(*args, **kwargs)
 
-    def post_training_step(self):
+    def post_training_step(self) -> None:
         pass
 
-    def validation_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
+    def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         """The actual validation step.
 
         See :meth:`~pytorch_lightning.core.module.LightningModule.validation_step` for more details
         """
         with self.precision_plugin.val_step_context():
+            assert isinstance(self.model, pl.LightningModule) or isinstance(
+                self.model, _LightningPrecisionModuleWrapperBase
+            )
             return self.model.validation_step(*args, **kwargs)
 
-    def test_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
+    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
         """The actual test step.
 
         See :meth:`~pytorch_lightning.core.module.LightningModule.test_step` for more details
         """
         with self.precision_plugin.test_step_context():
+            assert isinstance(self.model, pl.LightningModule) or isinstance(
```

Comment:
Hmm, not sure this would work with custom strategies who have their own wrappers. 
Shouldn't we check against a protocol here, e.g., a type expressing that the object has signatures for training_step, test_step, etc. to be able to call these methods? Essentially we want to rely on duck-typing here. Does that make sense?

Reply:
I think the idea was to implement something like 
```python
@runtime_checkable
class Trainable(Protocol):
    """Trainable protocol requires all ``*_step`` so that it can safely be used by ``Strategy`` classes."""
    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
        ...

    def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
        ...

    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
        ...

    def predict_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
        ...
```
However this is not flexible enough for the classes implementing only few of them such as [this test model](https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/tests/tests_pytorch/trainer/test_dataloaders.py#L1418) which does not implement ```validation_step```.
Therefore I split those functions in 4 different protocols. Even though it works I'm not sure it's clean.